### ISSUE TITLE
Core resource allocation using time-slice

### ISSUE THREAD

**Issue by ltson4121994** [Project-HAMi/HAMi#397]
Hi, I came across this repos from tkestack/gpu-manager. I saw that in gpu-manager, vcuda was used for core resource sharing, I wonder why it was changed to time-slicing. In my opinion, the ability to execute parallel would be more optimized compared to time slicing. Could you please share more about this design decision? Thank you very much.

**Comment by archlitchi** [comment]
If you want to use core resource sharing, you need to put all different tasks into the same context, such technics is called MPS, but there are many issues on MPS, so we can't apply that in production environment.

**Comment by ltson4121994** [comment]
Hi, thanks for your prompt reply. I think this project is still using hami core right? I take a look at the source code and it seems like the sharing method is by allocating cores for each kernel launch not the nvidia time slicing method right? Have you ever try benchmarking the performance?

**Comment by archlitchi** [comment]
yes, hami-core is not using nvidia time slicing method, it implements its own time slicing strategy by blocking kernel submission of certain tasks periodically. the performance introduced is less than 5%

**Comment by ltson4121994** [comment]
I mean the performance comparation between nvidia time slicing and the custom time slicing when there are multiple containers launch kernel at the same time. Theoretically, I think the custom method should be better since it allow parallel kernel launching while nvidia time slicing does not?

**Comment by archlitchi** [comment]
both nvidia and custom don't support parallel kernel launching

**Comment by archlitchi** [comment]
em... i'm not sure if we can use nvidia time slicing in HAMi-core,  we can discuss that if you have a plan

**Comment by ltson4121994** [comment]
I think for nvidia time slicing we only need to config the node but it is not as flexible since we have to specify the number of replicas and whenever this config is changed, gpu-operator has to be restarted so I guess it is not production ready as well. I am just curious about the performance comparison in practice. But if both cannot execute kernel launch in parallel, then probably nvidia time slicing will have better performance since all containers have full access to the resources. In that case, we are somewhat trading performance for better governance right?

**Comment by archlitchi** [comment]
i think i get what nvidia time-slicing means, it simply throw that task into GPU, and let them compete with each other. It doesn't guarantee how many compute power it can use. it equals to submit all vGPU tasks without specify nvidia.com/gpucores.

**Comment by ltson4121994** [comment]
I see, but I still haven't figured why we should limit the core utilization when we are not able to run in parallel anyway. It's either we let them run in parallel and enforce core utilization quotas or we let them run sequentially with the entire core resources right? In that case we should enforce the quotas by execution time spent for each process I suppose, not sure if it is possible though.

**Comment by thien-lm** [comment]
@lixd @archlitchi  it mean that if two completely different process run on GPU, a process restrict 40% GPU usage and the other be set 60%, GPU will never reach 100% usage? 
as for now, i see that i can just enable MPS to allow CUDA process to run on GPU, there were no conflict with HAMI core IMO  😄
